### PR TITLE
AF-2529: Allow users to embed dashboards

### PR DESCRIPTION
**Pull request by jesuino**
This PR contains feature AF-2529 and a fix for AF-2527.

AF-2529:

With this feature you should be able to access dashboards directly using

{dashbuilder_runtime_host}/?standalone=true&perspective={pageId}

Where pageId is the id of the page you want to access. This will allow users to embedded dashboads in HTML pages.

AF-2527:
Quick fix to allow imports without datasets (no exported datasets) or no navigation (cases where the navigation isn't created at all) to import their dashboards. The only required component is page.

General fixes:

Check style fixes a few more javadocs.

### CODE REVIEW

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by jesuino** [comment]
@sthundat @barboras7 Hello, do you know who could review this? Thanks!

**Comment by jesuino** [comment]
Jenkins execute full downstream build

File: dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/client/ClientRuntimeModelLoader.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        importModelServiceCaller.call((Optional<RuntimeModel> runtimeModelOp) -> {
+            if (runtimeModelOp.isPresent()) {
+                RuntimeModel runtimeModel = runtimeModelOp.get();
+                runtimeModel.getLayoutTemplates().forEach(perspectiveEditorGenerator::generatePerspective);
+                modelCache = runtimeModel;
+                modelLoaded.accept(runtimeModel);
+            } else {
+                emptyModel.execute();
+            }
+        }, (ErrorCallback<Exception>) (Exception message, Throwable throwable) -> {
```
Comment (caponetto) [inline dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/client/ClientRuntimeModelLoader.java:72]:
```suggestion
        }, (message, throwable) -> {
```

**Comment by sthundat** [comment]
@jesuino I will look into this
thanks,
Savitha

File: dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/shared/model/ImportDefinitions.java [hunk 2]
```diff
@@ -22,13 +22,15 @@
...
+    public static final String DATASET_PREFIX = "dashbuilder/datasets";
+    public static final String DATASET_DEF_PREFIX = DATASET_PREFIX + "/definitions";
+    public static final String DATASET_SUFFIX = ".dset";
+    
+    public static final String PERSPECTIVE_PREFIX = "dashbuilder/perspectives";
     public static final String PERSPECTIVE_SUFFIX = "perspective_layout";
 
-    public static final String NAVIGATION_FILE = NAVIGATION_PREFIX + "navtree.json";
+    public static final String NAVIGATION_PREFIX = "dashbuilder/navigation";
+    public static final String NAVIGATION_FILE = NAVIGATION_PREFIX + "/navigation/navtree.json";
```
[the remark this answers is not in the document]
Comment (jesuino) [reply dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/shared/model/ImportDefinitions.java:33]: Yes, it is stored this way:

image

**Comment by jesuino** [comment]
Hello @sthundat Thanks for looking into it. The main change is that now you should be able to embedded dashboards. To make it possible we enabled standalone mode to work with runtime perspectives, hence you should be able to open any page using the page name.

You can download the WAR from here:

https://drive.google.com/file/d/13EsYgXZY8ArC-MVLq2gy0HHtJu19i_7H/view?usp=sharing

It responds on root context (/) and only requires an admin user to login. Once you import a dashboard you should be able to open it in standalone mode using:

{host}?standalone=true&perspective={perspective}

For example, my page is life_expectancy, and this is what shows in the browser when I navigate to it:

http://localhost:8080/#life_expectancy%7Clife_expectancy%20%5BScreen%5D

The standalone page is:

http://localhost:8080?standalone=true&perspective=life_expectancy

This same URL could be used in iframe in HTML pages.

**Comment by sthundat** [comment]
@jesuino can you please add unit tests for ClientRuntimeModelLoader

**Comment by sthundat** [comment]
Hi William,
please find some comments below,
1.if i given some wrong page name in standalone mode it is redirect to a page as shown below,and ok button of popup dialog does not clickable.
activitynotfound.
2.Can you give a relevant error message instead of activity not found'

**Comment by jesuino** [comment]
Hello @sthundat 

Thanks again for your review.

I made a change in runtime, now it has a not found page for standalone (embedded) mode:

image

I also added tests for ClientRuntimeModelLoader. I will improve tests when Multi mode comes out, because multi will require internal changes to some classes. I will update this in a few minutes with the link to a new dashbuilder runtime war.

**Comment by sthundat** [comment]
@jesuino ,
new 404 page looks great!!...thanks for the fix

please find some more comments on new build below,

1.if i add a page with name contain url encoding characters for example 'test%5DTest' we are getting 404 page in embedded mode
standalone url:
http://localhost:8080?standalone=true&perspective=test%5DTest
404 replaced url:
http://localhost:8080/?standalone&perspective=NotFoundPerspective&_perspective=test]Test#NotFoundPerspective%7CNotFoundScreen
this may be a rare situation but I guess we can fix this as well.

2.if a user by mistake import a page without any data or datasets , then that page is redirect to a blank page in runtime dashboards.
Instead of blank page can we include some warning/message to make it more user friendly.
WDYT?
Thanks,
Savitha

**Comment by jesuino** [comment]
Hello Savitha,

Thanks for the review. Here are my answers:

1) In this case the page itself contains special chars, hence you will have to re-encode the URL:

standalone=true&perspective=test%255DTest

2) I do agree with you, a message saying that the page is completely empty would help (empty layout). However, I think this could be done in another JIRA, wdyt? Would you please create it? 

Thanks!

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by sthundat** [comment]
Hi @jesuino ,
Thanks for the inputs ,I have created a jira for the empty page message.
please find it [here](https://issues.redhat.com/browse/AF-2569)
Thanks,
Savitha

**Comment by jesuino** [comment]
thank you @sthundat really appreciate

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by jesuino** [comment]
The errors I see in logs does not seem to be related https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/KIE/job/master/job/upstream-downstream/job/appformer.downstream/143/

**Comment by jesuino** [comment]
The failing tests are not related:

~~~
Test Result (3 failures / +3)
org.kie.processmigration.it.ProcessMigrationIntegrationTest.testReadinessCheck
org.kie.processmigration.it.ProcessMigrationIntegrationTest.testMigration
org.kie.processmigration.it.ProcessMigrationIntegrationTest.testLivenessCheck
~~~
This is being discussed internally (see https://issues.redhat.com/browse/JBPM-9196 )

**Pull request merged**